include/deal.II/base/mpi.templates.h

```diff
@@ -186,6 +186,129 @@ namespace Utilities
               std::copy(values.begin(), values.end(), output.begin());
           }
       }
+
+      template<typename T>
+      std::map<unsigned int, std::vector<T> >
+      send_and_receive(const MPI_Comm                                         &comm,
+                       const std::map<unsigned int, typename std::vector<T> > &objects_to_send)
+      {
+#ifndef DEAL_II_WITH_MPI
+        (void)comm;
+        (void)objects_to_send;
+        Assert (false, ExcMessage ("The function send_and_receive doesn't make"
+                                   "any sense if you do not have MPI enabled. "));
+#else
+        auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+        auto my_proc = dealii::Utilities::MPI::this_mpi_process(comm);
+
+        // Collective communication telling each process
+        // how many objects I'm sending: this value shall be contained
+        // inside send_amount
+        std::vector<unsigned int> send_amount(n_procs, 0);
+        // Vector to store, in each entry, the number of objects the corresponding
+        // process shall send to this process
+        std::vector<unsigned int> rec_amount(n_procs, 0);
+
+        for (const auto &m: objects_to_send)
+          send_amount[m.first] = m.second.size();
+
+        MPI_Alltoall(
+          &send_amount[0], 1, MPI_UNSIGNED,
+          &rec_amount[0], 1, MPI_UNSIGNED, comm);
```

Comment:
I *think* that what you are doing here is to figure out who needs to get what. There is already a function to do that: Utilities::MPI::compute_point_to_point_communication_pattern`. It uses an `MPI_Allgather` instead of `MPI_Alltoall`, so should be more efficient.